src/Compiler/Driver/ReuseTcResults/ReuseTcResultsDriver.fs

```diff
@@ -0,0 +1,124 @@
+module internal FSharp.Compiler.ReuseTcResults
+
+open System
+open System.Collections.Generic
+open System.IO
+
+open FSharp.Compiler.CompilerConfig
+open FSharp.Compiler.Diagnostics
+open FSharp.Compiler.GraphChecking
+open FSharp.Compiler.IO
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.Syntax.PrettyNaming
+
+open Internal.Utilities.Hashing
+
+let (++) a b = Path.Combine(a, b)
+
+[<Sealed>]
+type ReuseTcResultsDriver(tcConfig: TcConfig) =
+
+    let outputDir = tcConfig.outputDir |> Option.defaultValue ""
+    let tcDataFolderName = outputDir ++ FSharpTcDataResourceName
+
+    let cmdLineFilePath = tcDataFolderName ++ "cmdline"
+    let graphFilePath = tcDataFolderName ++ "graph"
+    let referencesFilePath = tcDataFolderName ++ "references"
+
+    let writeThisTcData path content =
+        use tcDataFile = FileSystem.OpenFileForWriteShim path
+        tcDataFile.WriteAllLines content
+
+    let readPrevTcData path =
+        if FileSystem.FileExistsShim path then
+            use tcDataFile = FileSystem.OpenFileForReadShim path
+            tcDataFile.ReadAllLines()
+        else
+            Array.empty
+
+    let getPrevCompilationCmdLine () = readPrevTcData cmdLineFilePath
+    let getPrevCompilationGraph () = readPrevTcData graphFilePath
+    let getPrevCompilationReferences () = readPrevTcData referencesFilePath
+
+    let writeThisCompilationCmdLine = writeThisTcData cmdLineFilePath
+    let writeThisCompilationGraph = writeThisTcData graphFilePath
+    let writeThisCompilationReferences = writeThisTcData referencesFilePath
+
+    let getContentHash fileName =
+        use stream = FileSystem.OpenFileForReadShim fileName
+        stream.ReadAllBytes() |> Md5Hasher.computeHash |> BitConverter.ToString
+
+    let getThisCompilationCmdLine args = args
+
+    // maybe split into two things?
+    let getThisCompilationGraph inputs =
+        let sourceFiles =
+            inputs
+            |> Seq.toArray
+            |> Array.mapi (fun idx (input: ParsedInput) ->
+                {
+                    Idx = idx
+                    FileName = input.FileName
+                    ParsedInput = input
+                })
+
+        let filePairs = FilePairMap sourceFiles
+        let graph, _ = DependencyResolution.mkGraph filePairs sourceFiles
+
+        let list = List<string>()
+
+        for KeyValue(idx, _) in graph do
+            let hash = getContentHash sourceFiles[idx].FileName
+            list.Add($"%i{idx}[\"{hash}\"]")
+
+        for KeyValue(idx, deps) in graph do
+            for depIdx in deps do
+                list.Add($"%i{idx} --> %i{depIdx}")
+
+        list.ToArray()
+
+    let getThisCompilationReferences =
+        List.map (fun (r: AssemblyReference) -> r.Text, getContentHash r.Text)
+        >> List.map (fun (name, hash) -> $"{name}: {hash}")
+        >> List.toArray
```

Comment:
We can do those things. Checking hash helps in the scenarios where one changes a dependency, reverts it immediately (ctrl + Z) and then triggers the recompilation. If we only look at the date, this will look like a new recompilation.

We can different processes for system and non-system dependencies, do you think it's worth it?

Reply:
I think we should just straight up use MVID for assemblies. If it's good for deterministic builds and msbuild, it should be good enough for us. This way you'll only be reading header, and not calculate anything.

Reply:
Changed for using LastWriteTime for now, that's also a rather lightweight thing.